File: MgDev/UnitTest/WebTier/DotNet/TestCommon/TestExecutionRun.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using SqliteDotNet;

namespace OSGeo.MapGuide.Test.Common
{
    /// <summary>
    /// Executes a <see cref="T:OSGeo.MapGuide.Test.Common.ITestExecutorCollection"/> instance
    /// </summary>
    public class TestExecutionRun : IDisposable
    {
        private ITestExecutorCollection _executors;
        private SqliteDb _unitTestDb;
        private SqliteVm _unitTestVm;
        private SqliteVm _unitTestResultVm;
        private bool _generateDump;

        public TestExecutionRun(string dbPath, ITestExecutorCollection exec, bool generateDump = false)
        {
            _executors = exec;
            _generateDump = generateDump;
            _unitTestDb = new SqliteDb();
            _unitTestDb.Open(dbPath);

            _unitTestVm = new SqliteVm(_unitTestDb, false);
            _unitTestResultVm = new SqliteVm(_unitTestDb, false);
        }

        public void Dispose()
        {
            _unitTestResultVm.SqlFinalize();
            _unitTestResultVm = null;
            _unitTestVm.SqlFinalize();
            _unitTestVm = null;
            try
            {
                _unitTestDb.Close();
            }
            catch { }
            _unitTestDb = null;
        }

        public int RunTests(string testName, ITestLogger logger, ref int testsRun)
        {
            TestResult actualResult = null;
            int exitStatus = 0;
            try
            {
                //Add the test in the log file
                string msg = "\n\n************************************************************\n" +
                             string.Format("Executing {0}\n", testName) +
                             "************************************************************\n\n";

                logger.Write(msg);

                //Get the list of operations to be executed for this test. This list should comma separated, no spaces.
                _unitTestVm.Execute("Select ParamSets from TestCase where TestName=\"{0}\"", testName);
                string sets = _unitTestVm.GetString("ParamSets");

                //Extract the paramset ids
                int[] paramSet = sets.Split(',').Where(x => !string.IsNullOrEmpty(x)).Select(x => Convert.ToInt32(x)).ToArray();
                foreach (int paramSetId in paramSet)
                {
                    _unitTestVm.Execute("Select ParamValue from Params where ParamSet={0} AND ParamName=\"OPERATION\"", paramSetId);
                    string paramValue = _unitTestVm.GetString("ParamValue");

                    testsRun++;

                    //Add the operation to the log file
                    AddLogFileEntry(logger, paramValue, paramSetId, _unitTestResultVm);

                    var exec = _executors.GetTestExecutor(paramValue);
                    actualResult = exec.Execute(paramSetId);

                    exitStatus += _executors.ValidateRequest(_unitTestDb, testName, paramSetId, paramValue, actualResult, logger);
                }

                return exitStatus;
            }
            catch (MgException ex)
            {
                throw new UnitTestException(string.Format("Exception from MapGuide:\n{0}", ex.GetDetails()));
            }
        }

        private static void AddLogFileEntry(ITestLogger logger, string operation, int paramSetId, SqliteVm vm)
        {
            logger.Write("\nParamSet: {0}\n", paramSetId);
            int status = vm.Execute("Select * from Params where ParamSet={0}", paramSetId);

            while (status == Sqlite.Row)
            {
                string paramName = vm.GetString("ParamName");
                string paramValue = vm.GetString("ParamValue");

                logger.Write("{0}: {1}\n", paramName, paramValue);
                status = vm.NextRow();
            }
        }

        public void Cleanup()
        {
            if (_executors != null)
            {
                _executors.Cleanup();
                _executors = null;
            }
        }
    }
}

File: MgDev/UnitTest/WebTier/Java/src/org/osgeo/mapguide/test/common/TestExecutionRun.java
package org.osgeo.mapguide.test.common;

import java.io.File;
import com.almworks.sqlite4java.*;
import org.osgeo.mapguide.*;
import org.osgeo.mapguide.test.*;

public class TestExecutionRun implements java.io.Closeable
{
    private ITestExecutorCollection _executors;
    private SQLiteConnection _unitTestDb;

    public TestExecutionRun(String dbPath, ITestExecutorCollection exec) throws SQLiteException
    {
        _executors = exec;
        _unitTestDb = new SQLiteConnection(new File(dbPath));
        _unitTestDb.open(false);
    }

    public void close()
    {
        try
        {
            _unitTestDb.dispose();
        }
        catch (Exception ex) {

        }
        _unitTestDb = null;
    }

    private int[] ParamSetStrToArray(String pSet) {
        if (CommonUtility.IsNullOrEmpty(pSet))
            return new int[0];

        String[] parts = pSet.split(",");
        int[] result = new int[parts.length];
        int i = 0;
        for (String s : parts) {
            result[i] = Integer.parseInt(s.trim());
            i++;
        }
        return result;
    }

    public int RunTests(String testName, ITestLogger logger, MutableInteger testsRun) throws SQLiteException
    {
        TestResult actualResult = null;
        int exitStatus = 0;
        try
        {
            //Add the test in the log file
            String msg = "\n\n************************************************************\n" +
                         String.format("Executing %s\n", testName) +
                         "************************************************************\n\n";

            logger.Write(msg);

            //Get the list of operations to be executed for this test. This list should comma separated, no spaces.
            SQLiteStatement stmt = _unitTestDb.prepare(String.format("Select ParamSets from TestCase where TestName=\"%s\"", testName));
            while (stmt.step()) {
                String sets = CommonUtility.GetStringFromStatement(stmt, "ParamSets");

                //Extract the paramset ids
                int[] paramSet = ParamSetStrToArray(sets);
                for (int paramSetId : paramSet)
                {
                    SQLiteStatement pstmt = _unitTestDb.prepare(String.format("Select ParamValue from Params where ParamSet=%d AND ParamName=\"OPERATION\"", paramSetId));
                    while (pstmt.step()) {
                        String paramValue = CommonUtility.GetStringFromStatement(pstmt, "ParamValue");

                        //Add the operation to the log file
                        AddLogFileEntry(logger, paramValue, paramSetId);

                        try {
                            ITestExecutor exec = _executors.GetTestExecutor(paramValue);
                            if (DebugSettings.Enabled)
                                Console.WriteLine("\tExecuting operation: %s", exec.getOperationName());
                            actualResult = exec.Execute(paramSetId);
                            testsRun.increment();
                            exitStatus += _executors.ValidateRequest(_unitTestDb, testName, paramSetId, paramValue, actualResult, logger);
                        } catch (MissingTestExecutorException ex) {
                            if (DebugSettings.Enabled)
                                Console.WriteLine("No test executor for %s in test: %s. Skipping test.", paramValue, testName);
                        }
                    }
                }
            }
            return exitStatus;
        }
        catch (MgException ex)
        {
            throw new UnitTestException(String.format("Exception from MapGuide:\n%s", ex.getDetails()));
        }
    }

    private void AddLogFileEntry(ITestLogger logger, String operation, int paramSetId) throws SQLiteException
    {
        logger.Write("\nParamSet: %s\n", paramSetId);
        SQLiteStatement stmt = _unitTestDb.prepare(String.format("Select * from Params where ParamSet=%d", paramSetId));

        while (stmt.step())
        {
            String paramName = CommonUtility.GetStringFromStatement(stmt, "ParamName");
            String paramValue = CommonUtility.GetStringFromStatement(stmt, "ParamValue");

            logger.Write("%s: %s\n", paramName, paramValue);
        }
    }

    public void Cleanup()
    {
        if (_executors != null)
        {
            _executors.Cleanup();
            _executors = null;
        }
    }
}